### PR TITLE
Refactor connection setup to allow runtime transport registration

### PR DESCRIPTION
**Pull request by Hexxeh**
The intention of this is to allow tools that build upon pebble-tool tool to provide their own transport mechanisms without modifying this package to give it knowledge of them.

### CODE REVIEW

**Comment by astapleton** [comment]
You need to update the libpebble2 version requirement for this to work: libpebble2>=0.0.21

File: pebble_tool/commands/base.py [hunk 1]
```diff
@@ -70,37 +70,28 @@ def _set_debugging(self, level):
...
-            group.add_argument('--serial', type=str, help="Connected directly, given a path to a serial device.")
+        group = parser.add_mutually_exclusive_group()
+        for handler_impl in cls.valid_connection_handlers():
+            handler_impl.add_argument_handler(group)
         return super(PebbleCommand, cls)._shared_parser() + [parser]
 
+    @classmethod
+    def valid_connection_handlers(cls):
+        valid_connections = getattr(cls, 'valid_connections', None)
+        if valid_connections:
```
Comment (Katharine) [inline pebble_tool/commands/base.py]: Did you mean `if not valid_connections`?

Comment (Hexxeh) [inline pebble_tool/commands/base.py]: Welp, yes. Rejigged this slightly before uploading from using hasattr and screwed up. Fixed!

Comment (Katharine) [inline pebble_tool/commands/base.py]: The magic of code review!

File: pebble_tool/commands/base.py [hunk 2]
```diff
@@ -110,107 +101,186 @@ def __call__(self, args):
...
-                running = []
-                emulator_platform = None
-                emulator_sdk = None
-                if 'PEBBLE_EMULATOR' in os.environ:
-                    emulator_platform = os.environ['PEBBLE_EMULATOR']
-                    if emulator_platform not in pebble_platforms:
-                        raise ToolError("PEBBLE_EMULATOR is set to '{}', which is not a valid platform "
-                                        "(pick from {})".format(emulator_platform, ', '.join(pebble_platforms)))
-                    emulator_sdk = os.environ.get('PEBBLE_EMULATOR_VERSION', sdk_version())
-                else:
```
Comment (Katharine) [inline pebble_tool/commands/base.py]: You've removed the functionality implemented by this block: if exactly one emulator is running, and no connection is specified, we assume you meant that one.

Comment (Hexxeh) [inline pebble_tool/commands/base.py]: Re-implemented this behaviour.

**Comment by Hexxeh** [comment]
@Katharine anything else needed on this?

**Comment by Katharine** [comment]
Launching multiple emulators of different types is now broken; after the first one is launched it just ignores the type parameter and always connects to that.

The message complaining about multiple emulators when no connection is specified also seems to have gotten lost.

**Comment by Hexxeh** [comment]
The multiple emulator error appears okay for me?

```blackjack:pebble-tool Hexxeh$ ./pebble.py repl
Multiple emulators are running; you must specify which to use.```

Fixed the problem launching multiple emulators.

**Comment by Hexxeh** [comment]
Will need to push libpebble2 0.0.22 to pypi before landing this I think.

**Comment by Katharine** [comment]
The multiple emulator behaviour still isn't working when I test it.

**Comment by bradtgmurray** [comment]
screen shot 2016-07-21 at 1 41 47 pm

Tried to test this out, seems to be working?

**Comment by astapleton** [comment]
Can we get this landed?

**Pull request merged**